File: Problem2/problem2.js
"use strict"

/**
 * Adds up the even numbers in a fibonacci sequence until it reaches the upper limit.
 * 
 * @param {Number} upperLimit           Upper limit.
 * @returns {Number}                    The result of adding up the even numbers in the fibonacci sequence.
 */
let addEvenFibonacciNumbers = (upperLimit) => {
    
    /**
     * Function that generates and adds the even numbers in a fibonacci sequence.
     * 
     * @param {*} firstNumber           The N-1 number for generating the next fibonacci number.
     * @param {*} secondNumber          The N number for generating the next fibonacci number.
     * @param {*} index                 Index of the fibonacci sequence.
     * @param {*} upperLimit            Upper limit
     * @returns {Number}                The result of adding up the even numbers in the fibonacci sequence.
     */
    function generateFibonacciNums(firstNumber, secondNumber, index, upperLimit){
        let result = 0;
        if(index > upperLimit)
            return 0;
        else{
            result = generateFibonacciNums(secondNumber, (firstNumber + secondNumber), ++index, upperLimit);
            
            if((firstNumber + secondNumber) % 2 === 0)
                return (result + firstNumber + secondNumber);
            else   
                return result;
        }
    }

    return generateFibonacciNums(0, 1, 1, upperLimit);
}


console.log(addEvenFibonacciNumbers(10));
console.log(addEvenFibonacciNumbers(18));
console.log(addEvenFibonacciNumbers(23));
console.log(addEvenFibonacciNumbers(43));

File: Problem2/Problem2.java

import java.util.stream.Stream;

public class Problem2{

    public static void main(String args[]){
        System.out.println(addEvenFibonacciNumbers(10));
        System.out.println(addEvenFibonacciNumbers(18));
        System.out.println(addEvenFibonacciNumbers(23));
        System.out.println(addEvenFibonacciNumbers(43));
    }

    /**
     * It adds up the even numbers in a fibonacci sequence until an upper limit.
     * @param upperLimit            Upper limit.
     * @return                      The result of adding up the even numbers.
     */
    private static int addEvenFibonacciNumbers(int upperLimit){
        return Stream.iterate(new int[]{1,2}, n -> new int[]{n[1], n[0] + n[1]})
        .limit(upperLimit)
        .map(n -> n[0])
        .filter(n -> n % 2 == 0)
        .mapToInt(n -> n)
        .sum();   
    }
}